### DISCUSSION TITLE
Bug at render-markdown stage of tutorial:  Must use import to load ES Module

### DISCUSSION THREAD

**Discussion by danbarker** [vercel/next.js#27758] [Help] [answered]
I have been following the Next.js tutorial, and got to this step: https://nextjs.org/learn/basics/dynamic-routes/render-markdown

However, when trying to run the code as given, an error is returned:

```
Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /Users/danbarker/Dev/nextjs-blog/node_modules/remark/index.js
require() of ES modules is not supported.
require() of /Users/danbarker/Dev/nextjs-blog/node_modules/remark/index.js from /Users/danbarker/Dev/nextjs-blog/.next/server/pages/index.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
Instead rename /Users/danbarker/Dev/nextjs-blog/node_modules/remark/index.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /Users/danbarker/Dev/nextjs-blog/node_modules/remark/package.json.

    at new NodeError (node:internal/errors:370:5)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1112:13)
    at Module.load (node:internal/modules/cjs/loader:975:32)
    at Function.Module._load (node:internal/modules/cjs/loader:816:12)
    at Module.require (node:internal/modules/cjs/loader:999:19)
    at require (node:internal/modules/cjs/helpers:93:18)
    at Object.remark (/Users/danbarker/Dev/nextjs-blog/.next/server/pages/index.js:486:18)
    at __webpack_require__ (/Users/danbarker/Dev/nextjs-blog/.next/server/webpack-runtime.js:33:42)
    at eval (webpack-internal:///./lib/posts.js:10:64)
    at Object../lib/posts.js (/Users/danbarker/Dev/nextjs-blog/.next/server/pages/index.js:83:1) {
  code: 'ERR_REQUIRE_ESM'
```

I have been looking around and think that this is perhaps something to do with Next.js not supporting ESM, and remark now apparently being ESM only. Is this right? If so, does anyone have an idea of a way around this? I have seen various other similar issues, e.g.
https://github.com/vercel/next.js/discussions/21253

Is the intro tutorial really not working, or am I missing something?

**Comment by stefanprobst** [comment, marked as answer]
`remark` (and the `unified` ecosystem more broadly) is currently transitioning to ESM-only, yes. so you could try just using a previous version (`remark@13` should work i think) - or you can try if it works with setting:
```js
// next.config.js
const config = {
  experimental: {
    esmExternals: true,
  },
}
module.exports = config
```
(requires next@canary version)

**Reply by g4s8** [reply]
Thanks, I had the same issue, solved by downgrading remark to `13` and clearing `.next` cache (`rm -fr .next`):
```json
  "dependencies": {
    "classnames": "^2.3.1",
    "gray-matter": "^4.0.3",
    "next": "^11.0.0",
    "react": "17.0.2",
    "react-dom": "17.0.2",
    "remark": "13.0.0",
    "remark-html": "13.0.1"
  }
```

**Reply by wickahead** [reply]
Thank you for the help. As a total noob, it took me a minute or two to remember that I had to run `npm update` after modifying the package.json with the lower version number that you called out.

**Reply by BraydenTW** [reply]
Thanks for the solution @stefanprobst! Could using this experimental feature break my nextjs project potentially?

**Reply by leerob** [reply]
I would recommend using the previous version of remark if you have concerns about the experimental flag for ESM support 👍

**Reply by BraydenTW** [reply]
Ok, which version should I downgrade to?

**Reply by leerob** [reply]
`npm install remark@13 remark-html@13`

**Reply by BraydenTW** [reply]
Oh, sorry I meant to say that I am following the structure you used to create leerob.io and it is giving me the same error with the whole ESM not supported, but instead it is for `remark-gfm`.

![image](https://user-images.githubusercontent.com/47185402/129789408-9f826ff5-48e1-45cc-a4f7-c0a2e3091cda.png)
![image](https://user-images.githubusercontent.com/47185402/129789429-c7c42a0a-cab1-4623-b64f-cd4618e06f45.png)

**Reply by leerob** [reply]
I'm not following - I would recommending following the example in my repo if you do want to use that together. It's correctly working with the flag and latest versions.

**Reply by BraydenTW** [reply]
Ok, gotcha. I'll let the discussions tab here know if I have other questions along the way. Thanks!

**Comment by danbarker** [comment]
Thanks all, solved by specifying v13.0.0 of remark in package.json and then running npm update.

As a data point, this is likely to make us use Gatsby for our project instead, as it doesn't inspire a lot of confidence - I can see that this issue has been coming for ~2 years: https://github.com/vercel/next.js/issues/9607

**Reply by stefanprobst** [reply]
to be fair, `remark` was only updated 2 days ago (see https://github.com/remarkjs/remark/releases/tag/14.0.0)

and ESM support is actively being worked on (see [my comment above](https://github.com/vercel/next.js/discussions/27758#discussioncomment-1131084) about the experimental config)

**Reply by danbarker** [reply]
OK, fair - perhaps Gatsby will have other issues, too. It's just an unfortunate first impression!

**Comment by stefanprobst** [comment]
@leerob could you please adjust the tutorial [here](https://nextjs.org/learn/basics/dynamic-routes/render-markdown) to use pinned `remark` versions:

```diff
- npm install remark remark-html
+ npm install remark@13 remark-html@13
```

thanks!

**Reply by leerob** [reply]
This has been fixed, thank you!

**Reply by olets** [reply]
@leerob confirming the tutorial works with remark 14 if [their documentation](https://github.com/remarkjs/remark/releases/tag/14.0.0) is followed, i.e. the single `remark` export is imported. I also see that both those remark docs and the [remark-html docs](https://github.com/remarkjs/remark-html#use) prefer the convention of naming imported plugins `remark<Plugin>`. A patch with this change would support the latest remark and follow its conventions:


```diff
- npm install remark@13 remark-html@13
+ npm install remark remark-html
```
```diff
- import remark from 'remark'
- import html from 'remark-html'
+ import {remark} from 'remark'
+ import remarkHtml from 'remark-html'
```
```diff
  const processedContent = await remark()
-   .use(html)
+   .use(remarkHtml)
    .process(matterResult.content)
```

**Reply by leerob** [reply]
Created https://github.com/vercel/next-learn-starter/issues/45, thank you! Open for contributions.
